File: mobile_app/lib/services/sentry_service.dart
import 'dart:io';
import 'dart:async';
import 'package:flutter/foundation.dart';
import 'dart:developer' as dev;
import 'package:sentry_flutter/sentry_flutter.dart';
import 'package:device_info_plus/device_info_plus.dart';
import 'package:package_info_plus/package_info_plus.dart';
import 'package:connectivity_plus/connectivity_plus.dart';

/// Financial application specific error categories for Sentry
enum FinancialErrorCategory {
  authentication('authentication'),
  authorization('authorization'),
  transactionProcessing('transaction_processing'),
  paymentProcessing('payment_processing'),
  accountManagement('account_management'),
  budgetCalculation('budget_calculation'),
  financialAnalysis('financial_analysis'),
  dataValidation('data_validation'),
  externalIntegration('external_integration'),
  securityViolation('security_violation'),
  complianceIssue('compliance_issue'),
  uiError('ui_error'),
  networkError('network_error'),
  systemError('system_error');

  const FinancialErrorCategory(this.value);
  final String value;
}

/// Financial application specific severity levels
enum FinancialSeverity {
  low('low'),
  medium('medium'),
  high('high'),
  critical('critical'),
  securityCritical('security_critical');

  const FinancialSeverity(this.value);
  final String value;
}

/// Enhanced Sentry service for MITA Finance mobile application
class SentryFinancialService {
  static final SentryFinancialService _instance = SentryFinancialService._internal();
  factory SentryFinancialService() => _instance;
  SentryFinancialService._internal();

  bool _isInitialized = false;
  String? _userId;
  String? _userEmail; // Reserved for future user tracking
  String? _subscriptionTier; // Reserved for future tier-based monitoring
  String? _deviceInfo;
  
  /// Initialize comprehensive Sentry monitoring for financial application
  Future<void> initialize({
    required String dsn,
    String environment = 'development',
    String? release,
    bool enableCrashReporting = true,
    bool enablePerformanceMonitoring = true,
    bool enableUserInteractionTracing = true,
    double? sampleRate,
    double? tracesSampleRate,
  }) async {
    if (_isInitialized) {
      return;
    }

    try {
      // Get device and app information
      await _collectDeviceInfo();
      
      await SentryFlutter.init(
        (options) {
          options.dsn = dsn;
          options.environment = environment;
          options.release = release ?? 'mita-mobile@1.0.0';
          
          // Performance monitoring configuration
          options.tracesSampleRate = tracesSampleRate ?? _getTracesSampleRate(environment);
          options.enableAutoPerformanceTracing = enablePerformanceMonitoring;
          options.enableUserInteractionTracing = enableUserInteractionTracing;
          options.enableAutoSessionTracking = true;
          
          // Error sampling
          options.sampleRate = sampleRate ?? 1.0; // Capture all errors
          
          // Security and compliance for financial services
          options.sendDefaultPii = false; // Don't send PII data
          options.attachStacktrace = true;
          options.maxBreadcrumbs = 100;
          
          // Debug configuration
          options.debug = environment == 'development' && kDebugMode;
          
          // Custom configuration for mobile
          options.enableWindowMetricBreadcrumbs = true;
          options.enableNativeCrashHandling = enableCrashReporting;
          options.enableAutoNativeBreadcrumbs = true;
          
          
          // Configure screenshot attachment (disabled for financial compliance)
          options.attachScreenshot = false; // Disabled for financial privacy
          
          // Set before send callbacks
          options.beforeSend = _filterSensitiveData;
          options.beforeSendTransaction = (transaction) => _filterSensitiveTransactions(transaction, Hint());
        },
      );

      _isInitialized = true;
      
      // Configure initial scope after initialization
      await Sentry.configureScope((scope) {
        scope.setTag('application_type', 'financial_services');
        scope.setTag('compliance_level', 'pci_dss');
        scope.setTag('platform', Platform.operatingSystem);
        scope.setTag('environment', environment);
        scope.setTag('component', 'mobile_app');
        
        if (_deviceInfo != null) {
          scope.setTag('device_info', _deviceInfo!);
        }
        
        // Set application context
        scope.setContexts('application', {
          'name': 'MITA Finance Mobile',
          'version': '1.0.0',
          'type': 'financial_mobile_app',
          'compliance': 'PCI_DSS',
          'platform': Platform.operatingSystem,
        });
      });
      
      // Add initial breadcrumb
      Sentry.addBreadcrumb(Breadcrumb(
        message: 'Sentry financial monitoring initialized',
        category: 'init',
        level: SentryLevel.info,
        data: {
          'environment': environment,
          'financial_monitoring': true,
          'compliance_enabled': true,
        },
      ));
      
      if (kDebugMode) dev.log('Sentry financial monitoring initialized for $environment', name: 'SentryService');
      
    } catch (e, stackTrace) {
      if (kDebugMode) dev.log('Failed to initialize Sentry: $e', name: 'SentryService', error: e);
      if (kDebugMode) {
        if (kDebugMode) dev.log('Stack trace: $stackTrace', name: 'SentryService');
      }
      // Don't let Sentry initialization failure crash the app
    }
  }

  /// Get appropriate traces sample rate based on environment
  double _getTracesSampleRate(String environment) {
    switch (environment) {
      case 'production':
        return 0.1; // 10% for production
      case 'staging':
        return 0.5; // 50% for staging
      default:
        return 1.0; // 100% for development
    }
  }

  /// Collect device information for context
  Future<void> _collectDeviceInfo() async {
    try {
      final deviceInfo = DeviceInfoPlugin();
      final packageInfo = await PackageInfo.fromPlatform();
      
      String deviceDetails = '';
      
      if (Platform.isIOS) {
        final iosInfo = await deviceInfo.iosInfo;
        deviceDetails = '${iosInfo.name} ${iosInfo.systemVersion}';
      } else if (Platform.isAndroid) {
        final androidInfo = await deviceInfo.androidInfo;
        deviceDetails = '${androidInfo.brand} ${androidInfo.model} API${androidInfo.version.sdkInt}';
      }
      
      _deviceInfo = '${packageInfo.appName} v${packageInfo.version} on $deviceDetails';
      
    } catch (e) {
      _deviceInfo = 'Unknown device';
    }
  }

  /// Filter sensitive financial data before sending to Sentry
  SentryEvent? _filterSensitiveData(SentryEvent event, Hint hint) {
    // List of sensitive keys to redact
    const sensitiveKeys = {
      'password', 'token', 'secret', 'key', 'authorization',
      'card_number', 'cvv', 'pin', 'ssn', 'tax_id',
      'account_number', 'routing_number', 'sort_code',
      'iban', 'swift', 'bank_account', 'credit_card',
      'debit_card', 'payment_method', 'financial_data'
    };

    /// Recursively sanitize data structures
    dynamic sanitizeData(dynamic data) {
      if (data is Map) {
        final sanitized = <String, dynamic>{};
        for (final entry in data.entries) {
          final key = entry.key.toString().toLowerCase();
          if (sensitiveKeys.any((sensitiveKey) => key.contains(sensitiveKey))) {
            sanitized[entry.key as String] = '[REDACTED]';
          } else {
            sanitized[entry.key as String] = sanitizeData(entry.value);
          }
        }
        return sanitized;
      } else if (data is List) {
        return data.map(sanitizeData).toList();
      } else if (data is String && data.length > 1000) {
        return data.substring(0, 1000) + '... [TRUNCATED]';
      }
      return data;
    }

    // Sanitize request data
    if (event.request?.data != null) {
      event = event.copyWith(
        request: event.request!.copyWith(
          data: sanitizeData(event.request!.data),
        ),
      );
    }

    // Sanitize extra data
    if (event.extra?.isNotEmpty == true) {
      event = event.copyWith(
        extra: sanitizeData(event.extra) as Map<String, dynamic>?,
      );
    }

    // Add financial service tags
    final tags = Map<String, String>.from(event.tags ?? {});
    tags['financial_service'] = 'true';
    tags['pci_compliant'] = 'true';
    tags['data_sanitized'] = 'true';

    return event.copyWith(tags: tags);
  }

  /// Filter sensitive data from transaction events
  SentryTransaction? _filterSensitiveTransactions(SentryTransaction transaction, Hint hint) {
    // Add financial context to transactions
    final contexts = Map<String, dynamic>.from(transaction.contexts ?? {});
    contexts['financial_operation'] = {
      'type': 'financial_mobile_transaction',
      'compliance_level': 'pci_dss',
      'data_classification': 'confidential',
      'platform': Platform.operatingSystem,
    };

    return transaction.copyWith(contexts: Contexts.fromJson(contexts));
  }

  /// Capture financial-specific errors with enhanced context
  Future<SentryId> captureFinancialError(
    dynamic exception, {
    FinancialErrorCategory category = FinancialErrorCategory.systemError,
    FinancialSeverity severity = FinancialSeverity.medium,
    String? stackTrace,
    String? userId,
    String? transactionId,
    double? amount,
    String? currency,
    String? screenName,
    Map<String, dynamic>? additionalContext,
    Map<String, String>? tags,
  }) async {
    if (!_isInitialized) {
      if (kDebugMode) dev.log('Sentry not initialized - error not captured: $exception', name: 'SentryService');
      return const SentryId.empty();
    }

    return await Sentry.captureException(
      exception,
      stackTrace: stackTrace,
      withScope: (scope) {
        // Set financial error context
        scope.setTag('error_category', category.value);
        scope.setTag('severity', severity.value);
        scope.setTag('financial_error', 'true');
        
        // Set user context
        if (userId != null) {
          scope.setUser(SentryUser(id: userId));
        } else if (_userId != null) {
          scope.setUser(SentryUser(id: _userId));
        }

        // Set screen context
        if (screenName != null) {
          scope.setTag('screen_name', screenName);
          scope.setContexts('screen', {'name': screenName});
        }

        // Set financial context
        final financialContext = <String, dynamic>{
          'category': category.value,
          'severity': severity.value,
          'timestamp': DateTime.now().toIso8601String(),
          'platform': Platform.operatingSystem,
        };

        if (transactionId != null) {
          financialContext['transaction_id'] = transactionId;
        }
        if (amount != null) {
          financialContext['amount'] = amount;
        }
        if (currency != null) {
          financialContext['currency'] = currency;
        }

        scope.setContexts('financial_operation', financialContext);

        // Set additional context
        if (additionalContext != null) {
          for (final entry in additionalContext.entries) {
            scope.setExtra(entry.key, entry.value);
          }
        }

        // Set additional tags
        if (tags != null) {
          for (final entry in tags.entries) {
            scope.setTag(entry.key, entry.value);
          }
        }

        // Add network context
        _addNetworkContext(scope);
      },
    );
  }

  /// Capture transaction-specific errors
  Future<SentryId> captureTransactionError(
    dynamic exception, {
    required String userId,
    required String transactionId,
    required double amount,
    required String currency,
    required String transactionType,
    String? merchant,
    String? categoryName,
    String? screenName,
    String? stackTrace,
  }) {
    return captureFinancialError(
      exception,
      category: FinancialErrorCategory.transactionProcessing,
      severity: FinancialSeverity.high,
      stackTrace: stackTrace,
      userId: userId,
      transactionId: transactionId,
      amount: amount,
      currency: currency,
      screenName: screenName,
      additionalContext: {
        'transaction_type': transactionType,
        'merchant': merchant,
        'transaction_category': categoryName,
        'compliance_level': 'PCI_DSS',
        'financial_impact': 'DIRECT',
      },
    );
  }

  /// Capture authentication errors
  Future<SentryId> captureAuthenticationError(
    dynamic exception, {
    String? userId,
    String? authMethod,
    String? screenName,
    String? stackTrace,
    Map<String, dynamic>? additionalContext,
  }) {
    final context = <String, dynamic>{
      'auth_method': authMethod,
      'security_event': true,
      'compliance_impact': 'HIGH',
      'platform': Platform.operatingSystem,
      ...?additionalContext,
    };

    return captureFinancialError(
      exception,
      category: FinancialErrorCategory.authentication,
      severity: FinancialSeverity.securityCritical,
      stackTrace: stackTrace,
      userId: userId,
      screenName: screenName,
      additionalContext: context,
      tags: {
        'security_critical': 'true',
        'auth_failure': 'true',
      },
    );
  }

  /// Capture payment processing errors
  Future<SentryId> capturePaymentError(
    dynamic exception, {
    required String userId,
    required String paymentMethod,
    required double amount,
    required String currency,
    String? paymentProvider,
    String? screenName,
    String? stackTrace,
  }) {
    return captureFinancialError(
      exception,
      category: FinancialErrorCategory.paymentProcessing,
      severity: FinancialSeverity.critical,
      stackTrace: stackTrace,
      userId: userId,
      amount: amount,
      currency: currency,
      screenName: screenName,
      additionalContext: {
        'payment_method': paymentMethod,
        'payment_provider': paymentProvider,
        'pci_compliance_required': true,
        'financial_impact': 'DIRECT',
        'platform': Platform.operatingSystem,
      },
      tags: {
        'payment_error': 'true',
        'financial_critical': 'true',
      },
    );
  }

  /// Capture network errors with context
  Future<SentryId> captureNetworkError(
    dynamic exception, {
    String? endpoint,
    String? method,
    int? statusCode,
    String? screenName,
    String? stackTrace,
  }) async {
    final results = await Connectivity().checkConnectivity();
    final connectivity = results.isNotEmpty ? results.first : ConnectivityResult.none;
    final connectivityName = connectivity.name;
    
    return captureFinancialError(
      exception,
      category: FinancialErrorCategory.networkError,
      severity: FinancialSeverity.medium,
      stackTrace: stackTrace,
      screenName: screenName,
      additionalContext: {
        'endpoint': endpoint,
        'method': method,
        'status_code': statusCode,
        'connectivity': connectivityName,
        'platform': Platform.operatingSystem,
      },
      tags: {
        'network_error': 'true',
        'connectivity': connectivityName,
      },
    );
  }

  /// Set user context for financial application
  void setFinancialUser({
    required String userId,
    String? email,
    String? subscriptionTier,
    String? accountType,
    String? riskLevel,
  }) {
    if (!_isInitialized) return;

    _userId = userId;
    _userEmail = email;
    _subscriptionTier = subscriptionTier;

    Sentry.configureScope((scope) {
      scope.setUser(SentryUser(
        id: userId,
        email: email,
        data: {
          'subscription_tier': subscriptionTier,
          'account_type': accountType,
          'risk_level': riskLevel,
          'platform': Platform.operatingSystem,
        },
      ));

      // Set user-related tags
      scope.setTag('has_user_context', 'true');
      if (subscriptionTier != null) {
        scope.setTag('subscription_tier', subscriptionTier);
      }
      if (accountType != null) {
        scope.setTag('account_type', accountType);
      }
      if (riskLevel != null) {
        scope.setTag('risk_level', riskLevel);
      }
    });
  }

  /// Add financial breadcrumb
  void addFinancialBreadcrumb({
    required String message,
    String category = 'financial',
    SentryLevel level = SentryLevel.info,
    Map<String, dynamic>? data,
  }) {
    if (!_isInitialized) return;

    final breadcrumbData = <String, dynamic>{
      'timestamp': DateTime.now().toIso8601String(),
      'financial_context': true,
      'platform': Platform.operatingSystem,
      ...?data,
    };

    Sentry.addBreadcrumb(Breadcrumb(
      message: message,
      category: category,
      level: level,
      data: breadcrumbData,
    ));
  }

  /// Start performance transaction
  ISentrySpan startTransaction({
    required String name,
    required String operation,
    String? description,
    Map<String, dynamic>? data,
  }) {
    if (!_isInitialized) {
      return NoOpSentrySpan();
    }

    final transaction = Sentry.startTransaction(
      name,
      operation,
      description: description,
    );

    // Set financial context
    transaction.setTag('financial_operation', 'true');
    transaction.setTag('platform', Platform.operatingSystem);

    if (data != null) {
      for (final entry in data.entries) {
        transaction.setData(entry.key, entry.value);
      }
    }

    return transaction;
  }

  /// Add network context to scope
  void _addNetworkContext(Scope scope) {
    Connectivity().checkConnectivity().then((results) {
      final connectivity = results.isNotEmpty ? results.first : ConnectivityResult.none;
      final connectivityName = connectivity.name;
      scope.setContexts('network', {
        'connectivity': connectivityName,
        'platform': Platform.operatingSystem,
        'timestamp': DateTime.now().toIso8601String(),
      });
      scope.setTag('connectivity', connectivityName);
    }).catchError((e) {
      // Ignore network context errors
    });
  }

  /// Clear user context (e.g., on logout)
  void clearUser() {
    if (!_isInitialized) return;

    _userId = null;
    _userEmail = null;
    _subscriptionTier = null;

    Sentry.configureScope((scope) {
      scope.clearBreadcrumbs();
      scope.removeTag('subscription_tier');
      scope.removeTag('account_type');
      scope.removeTag('risk_level');
      scope.setTag('has_user_context', 'false');
      scope.setUser(null);
    });

    addFinancialBreadcrumb(
      message: 'User context cleared',
      category: 'auth',
      level: SentryLevel.info,
    );
  }

  /// Check if Sentry is initialized
  bool get isInitialized => _isInitialized;
}

/// Global instance
final sentryService = SentryFinancialService();

/// No-op Sentry span for when Sentry is not initialized
class NoOpSentrySpan implements ISentrySpan {
  @override
  Future<void> finish({SpanStatus? status, DateTime? endTimestamp}) async {}

  @override
  ISentrySpan startChild(String operation, {String? description, DateTime? startTimestamp}) => this;

  @override
  void setData(String key, dynamic value) {}

  @override
  void removeData(String key) {}

  @override
  void setTag(String key, String value) {}

  @override
  void removeTag(String key) {}

  @override
  SentrySpanContext get context => SentrySpanContext(operation: 'noop');

  @override
  DateTime get startTimestamp => DateTime.now();

  @override
  DateTime? get endTimestamp => DateTime.now();

  @override
  bool get finished => true;

  @override
  Map<String, String> get tags => {};

  @override
  Map<String, dynamic> get data => {};

  @override
  SpanStatus? get status => SpanStatus.ok();

  @override
  String? get description => null;

  @override
  String get origin => 'manual';

  @override
  SentryId get spanId => const SentryId.empty();

  @override
  SentryId? get parentSpanId => null;

  @override
  SentryId get traceId => const SentryId.empty();

  @override
  void setStatus(SpanStatus status) {}

  @override
  dynamic get localMetricsAggregator => null;

  @override
  void setMeasurement(String name, num value, {SentryMeasurementUnit? unit}) {}

  // Missing interface implementations
  @override
  set origin(String? origin) {}

  @override
  SentryTracesSamplingDecision? get samplingDecision => null;

  @override
  void scheduleFinish() {}

  @override
  set status(SpanStatus? status) {}

  @override
  dynamic get throwable => null;

  @override
  set throwable(dynamic throwable) {}

  @override
  SentryBaggageHeader? toBaggageHeader() => null;

  @override
  SentryTraceHeader toSentryTrace() => SentryTraceHeader(traceId, SpanId.fromId(spanId.toString()));

  @override
  SentryTraceContextHeader? traceContext() => null;
}

File: app/services/sentry_service.py
"""
Comprehensive Sentry Integration Service for MITA Finance
Provides advanced error monitoring, performance tracking, and financial-specific logging
"""

import logging
from datetime import datetime
from enum import Enum
from typing import Dict, Any, Optional
from contextlib import contextmanager

import sentry_sdk
from sentry_sdk import (
    set_user, set_tag, add_breadcrumb,
    capture_exception, capture_message, push_scope,
    start_transaction, start_span
)
from fastapi import Request

logger = logging.getLogger(__name__)


class FinancialErrorCategory(Enum):
    """Financial application specific error categories"""
    AUTHENTICATION = "authentication"
    AUTHORIZATION = "authorization"
    TRANSACTION_PROCESSING = "transaction_processing"
    PAYMENT_PROCESSING = "payment_processing"
    ACCOUNT_MANAGEMENT = "account_management"
    BUDGET_CALCULATION = "budget_calculation"
    FINANCIAL_ANALYSIS = "financial_analysis"
    DATA_VALIDATION = "data_validation"
    EXTERNAL_INTEGRATION = "external_integration"
    SECURITY_VIOLATION = "security_violation"
    COMPLIANCE_ISSUE = "compliance_issue"
    SYSTEM_ERROR = "system_error"


class FinancialSeverity(Enum):
    """Financial application specific severity levels"""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"
    SECURITY_CRITICAL = "security_critical"


class SentryFinancialService:
    """Enhanced Sentry service for financial applications"""
    
    def __init__(self):
        self.initialized = bool(sentry_sdk.Hub.current.client)
        if not self.initialized:
            logger.warning("Sentry not initialized - error monitoring will be limited")
    
    def capture_financial_error(
        self,
        exception: Exception,
        category: FinancialErrorCategory,
        severity: FinancialSeverity = FinancialSeverity.MEDIUM,
        user_id: Optional[str] = None,
        transaction_id: Optional[str] = None,
        amount: Optional[float] = None,
        currency: Optional[str] = None,
        request: Optional[Request] = None,
        additional_context: Optional[Dict[str, Any]] = None
    ) -> str:
        """
        Capture financial-specific errors with enhanced context
        
        Returns:
            str: Event ID from Sentry
        """
        if not self.initialized:
            logger.error(f"Financial error (not sent to Sentry): {exception}")
            return ""
        
        with push_scope() as scope:
            # Set financial error context
            scope.set_tag("error_category", category.value)
            scope.set_tag("severity", severity.value)
            scope.set_tag("financial_error", True)
            
            # Set user context
            if user_id:
                scope.set_user({"id": user_id})
            
            # Set financial context
            financial_context = {
                "category": category.value,
                "severity": severity.value,
                "timestamp": datetime.utcnow().isoformat()
            }
            
            if transaction_id:
                financial_context["transaction_id"] = transaction_id
            if amount is not None:
                financial_context["amount"] = amount
            if currency:
                financial_context["currency"] = currency
            
            scope.set_context("financial_operation", financial_context)
            
            # Set request context
            if request:
                self._set_request_context(scope, request)
            
            # Set additional context
            if additional_context:
                for key, value in additional_context.items():
                    scope.set_extra(key, value)
            
            # Add breadcrumb for financial operation
            add_breadcrumb(
                message=f"Financial error: {category.value}",
                category="financial",
                level="error",
                data={
                    "category": category.value,
                    "severity": severity.value,
                    "user_id": user_id,
                    "transaction_id": transaction_id
                }
            )
            
            # Capture the exception
            event_id = capture_exception(exception)
            
            # Log to application logger as well
            logger.error(
                f"Financial error captured [Sentry: {event_id}] - "
                f"Category: {category.value}, Severity: {severity.value}, "
                f"User: {user_id}, Transaction: {transaction_id}, Error: {str(exception)}"
            )
            
            return event_id or ""
    
    def capture_transaction_error(
        self,
        exception: Exception,
        user_id: str,
        transaction_id: str,
        amount: float,
        currency: str,
        transaction_type: str,
        merchant: Optional[str] = None,
        category: Optional[str] = None,
        request: Optional[Request] = None
    ) -> str:
        """Capture transaction-specific errors"""
        
        additional_context = {
            "transaction_type": transaction_type,
            "merchant": merchant,
            "transaction_category": category,
            "compliance_level": "PCI_DSS"
        }
        
        return self.capture_financial_error(
            exception=exception,
            category=FinancialErrorCategory.TRANSACTION_PROCESSING,
            severity=FinancialSeverity.HIGH,
            user_id=user_id,
            transaction_id=transaction_id,
            amount=amount,
            currency=currency,
            request=request,
            additional_context=additional_context
        )
    
    def capture_authentication_error(
        self,
        exception: Exception,
        user_id: Optional[str] = None,
        auth_method: Optional[str] = None,
        ip_address: Optional[str] = None,
        request: Optional[Request] = None
    ) -> str:
        """Capture authentication-specific errors"""
        
        additional_context = {
            "auth_method": auth_method,
            "ip_address": ip_address,
            "security_event": True,
            "compliance_impact": "HIGH"
        }
        
        return self.capture_financial_error(
            exception=exception,
            category=FinancialErrorCategory.AUTHENTICATION,
            severity=FinancialSeverity.SECURITY_CRITICAL,
            user_id=user_id,
            request=request,
            additional_context=additional_context
        )
    
    def capture_payment_error(
        self,
        exception: Exception,
        user_id: str,
        payment_method: str,
        amount: float,
        currency: str,
        payment_provider: str,
        request: Optional[Request] = None
    ) -> str:
        """Capture payment processing errors"""
        
        additional_context = {
            "payment_method": payment_method,
            "payment_provider": payment_provider,
            "pci_compliance_required": True,
            "financial_impact": "DIRECT"
        }
        
        return self.capture_financial_error(
            exception=exception,
            category=FinancialErrorCategory.PAYMENT_PROCESSING,
            severity=FinancialSeverity.CRITICAL,
            user_id=user_id,
            amount=amount,
            currency=currency,
            request=request,
            additional_context=additional_context
        )
    
    def capture_budget_error(
        self,
        exception: Exception,
        user_id: str,
        budget_category: str,
        calculation_type: str,
        request: Optional[Request] = None
    ) -> str:
        """Capture budget calculation errors"""
        
        additional_context = {
            "budget_category": budget_category,
            "calculation_type": calculation_type,
            "financial_accuracy_impact": True
        }
        
        return self.capture_financial_error(
            exception=exception,
            category=FinancialErrorCategory.BUDGET_CALCULATION,
            severity=FinancialSeverity.HIGH,
            user_id=user_id,
            request=request,
            additional_context=additional_context
        )
    
    @contextmanager
    def performance_monitoring(
        self,
        operation_name: str,
        operation_type: str = "financial_operation",
        user_id: Optional[str] = None,
        transaction_data: Optional[Dict[str, Any]] = None
    ):
        """Context manager for performance monitoring of financial operations"""
        
        if not self.initialized:
            # Provide a no-op context manager if Sentry is not initialized
            class NoOpContext:
                def __enter__(self):
                    return self
                def __exit__(self, *args):
                    pass
                def set_tag(self, key, value):
                    pass
                def set_data(self, key, value):
                    pass
            
            yield NoOpContext()
            return
        
        with start_transaction(
            op=operation_type,
            name=operation_name,
            sampled=True
        ) as transaction:
            
            # Set transaction context
            transaction.set_tag("operation_type", operation_type)
            transaction.set_tag("financial_operation", True)
            
            if user_id:
                transaction.set_tag("user_id", user_id)
            
            if transaction_data:
                for key, value in transaction_data.items():
                    transaction.set_data(key, value)
            
            # Add breadcrumb
            add_breadcrumb(
                message=f"Starting {operation_name}",
                category="performance",
                level="info",
                data={"operation": operation_name, "type": operation_type}
            )
            
            yield transaction
    
    @contextmanager
    def span_monitoring(
        self,
        span_name: str,
        description: Optional[str] = None,
        operation: str = "financial_span"
    ):
        """Context manager for span monitoring within transactions"""
        
        if not self.initialized:
            class NoOpSpan:
                def __enter__(self):
                    return self
                def __exit__(self, *args):
                    pass
                def set_tag(self, key, value):
                    pass
                def set_data(self, key, value):
                    pass
            
            yield NoOpSpan()
            return
        
        with start_span(
            op=operation,
            description=description or span_name
        ) as span:
            span.set_tag("span_name", span_name)
            yield span
    
    def add_financial_breadcrumb(
        self,
        message: str,
        category: str = "financial",
        level: str = "info",
        data: Optional[Dict[str, Any]] = None
    ):
        """Add financial-specific breadcrumb"""
        
        if not self.initialized:
            return
        
        breadcrumb_data = data or {}
        breadcrumb_data.update({
            "timestamp": datetime.utcnow().isoformat(),
            "financial_context": True
        })
        
        add_breadcrumb(
            message=message,
            category=category,
            level=level,
            data=breadcrumb_data
        )
    
    def set_financial_user_context(
        self,
        user_id: str,
        user_email: Optional[str] = None,
        subscription_tier: Optional[str] = None,
        account_type: Optional[str] = None,
        risk_level: Optional[str] = None
    ):
        """Set comprehensive user context for financial applications"""
        
        if not self.initialized:
            return
        
        user_context = {"id": user_id}
        
        if user_email:
            user_context["email"] = user_email
        if subscription_tier:
            user_context["subscription_tier"] = subscription_tier
        if account_type:
            user_context["account_type"] = account_type
        if risk_level:
            user_context["risk_level"] = risk_level
        
        set_user(user_context)
        
        # Set additional financial tags
        set_tag("has_user_context", True)
        if subscription_tier:
            set_tag("subscription_tier", subscription_tier)
        if account_type:
            set_tag("account_type", account_type)
        if risk_level:
            set_tag("risk_level", risk_level)
    
    def _set_request_context(self, scope, request: Request):
        """Set comprehensive request context"""
        
        # Basic request info (filtered for security)
        request_context = {
            "url": str(request.url.replace(query=None)),  # Remove query params
            "method": request.method,
            "endpoint": request.url.path,
            "user_agent": request.headers.get("user-agent", "")[:100],  # Truncate
            "content_type": request.headers.get("content-type", ""),
            "content_length": request.headers.get("content-length", "0")
        }
        
        # Add client IP (if available)
        if "x-forwarded-for" in request.headers:
            request_context["client_ip"] = request.headers["x-forwarded-for"].split(",")[0].strip()
        elif "x-real-ip" in request.headers:
            request_context["client_ip"] = request.headers["x-real-ip"]
        
        scope.set_context("request", request_context)
        
        # Set request tags
        scope.set_tag("endpoint", request.url.path)
        scope.set_tag("method", request.method)
        scope.set_tag("has_request_context", True)
    
    def capture_performance_issue(
        self,
        operation_name: str,
        duration_ms: float,
        threshold_ms: float,
        user_id: Optional[str] = None,
        additional_context: Optional[Dict[str, Any]] = None
    ) -> str:
        """Capture performance issues that exceed thresholds"""
        
        if not self.initialized:
            return ""
        
        with push_scope() as scope:
            scope.set_tag("performance_issue", True)
            scope.set_tag("operation", operation_name)
            scope.set_tag("exceeded_threshold", True)
            
            if user_id:
                scope.set_user({"id": user_id})
            
            # Performance context
            perf_context = {
                "operation": operation_name,
                "duration_ms": duration_ms,
                "threshold_ms": threshold_ms,
                "slowdown_factor": duration_ms / threshold_ms,
                "timestamp": datetime.utcnow().isoformat()
            }
            
            scope.set_context("performance", perf_context)
            
            if additional_context:
                for key, value in additional_context.items():
                    scope.set_extra(key, value)
            
            # Create a message for the performance issue
            message = (
                f"Performance threshold exceeded: {operation_name} "
                f"took {duration_ms:.2f}ms (threshold: {threshold_ms:.2f}ms)"
            )
            
            event_id = capture_message(message, level="warning")
            
            logger.warning(
                f"Performance issue captured [Sentry: {event_id}] - "
                f"{message}, User: {user_id}"
            )
            
            return event_id or ""


# Global instance
sentry_service = SentryFinancialService()


# Convenience functions for common use cases
def capture_financial_error(exception: Exception, **kwargs) -> str:
    """Convenience function to capture financial errors"""
    return sentry_service.capture_financial_error(exception, **kwargs)


def capture_transaction_error(exception: Exception, **kwargs) -> str:
    """Convenience function to capture transaction errors"""
    return sentry_service.capture_transaction_error(exception, **kwargs)


def capture_authentication_error(exception: Exception, **kwargs) -> str:
    """Convenience function to capture authentication errors"""
    return sentry_service.capture_authentication_error(exception, **kwargs)


def capture_payment_error(exception: Exception, **kwargs) -> str:
    """Convenience function to capture payment errors"""
    return sentry_service.capture_payment_error(exception, **kwargs)


def monitor_financial_performance(operation_name: str, **kwargs):
    """Convenience function for performance monitoring"""
    return sentry_service.performance_monitoring(operation_name, **kwargs)


def add_financial_breadcrumb(message: str, **kwargs):
    """Convenience function to add financial breadcrumbs"""
    return sentry_service.add_financial_breadcrumb(message, **kwargs)


def set_financial_user(user_id: str, **kwargs):
    """Convenience function to set financial user context"""
    return sentry_service.set_financial_user_context(user_id, **kwargs)